pkg/buildqueue/buildqueue.go

```diff
@@ -0,0 +1,568 @@
+//   chance to run but preemptive builds always preempt any queued, not yet
+//   running regular build.
+func (q *BuildQueue) Enqueue(b Builder) BuildNotification {
+	notify, enqueue := q.enqueueBuild(b)
+	if enqueue {
+		// This should be non-blocking unless there is contention beyond
+		// queueBuildSize in which case this will block until a slot in the
+		// queue becomes available.
+		q.workerBuildQueue <- b
+	}
+
+	return notify
+}
+
+// Remove removes the builder from the queue.
+func (q *BuildQueue) Remove(b Buildable) {
+	uuid := b.GetUUID()
+	q.mutex.Lock()
+	if status, ok := q.buildStatus[uuid]; ok {
+		delete(q.buildStatus, uuid)
+
+		if status.nextBuild != nil {
+			num := status.nextBuild.cancelScheduled(q)
+			defer status.builder.BuildsDequeued(num, true)
+		}
+	}
+	q.mutex.Unlock()
+}
+
+// Drain will drain the queue from any running or scheduled builds of a
+// Buildable. If a build is running, the function will block for the build to
+// complete. It is the responsibility of the caller that the Buildable does not
+// get re-scheduled during the draining. Returns true if waiting was required.
+func (q *BuildQueue) Drain(b Buildable) bool {
+	uuid := b.GetUUID()
+
+	q.mutex.Lock()
+	status, ok := q.buildStatus[uuid]
+	var currentBuild *build
+	if ok {
+		currentBuild = status.currentBuild
+		delete(q.buildStatus, uuid)
+
+		if status.nextBuild != nil {
+			num := status.nextBuild.cancelScheduled(q)
+			defer status.builder.BuildsDequeued(num, true)
+		}
+	}
+	q.mutex.Unlock()
+
+	// If a build is onging, block until build is complete
+	if currentBuild != nil {
+		<-currentBuild.internalNotification
+		return true
+	}
+
+	return false
+}
+
+func (q *BuildQueue) build(b Builder) error {
+	metric := metrics.BuildQueueEntries.With(map[string]string{
+		metrics.LabelBuildQueueName:  q.name,
+		metrics.LabelBuildQueueState: metrics.BuildQueueRunning,
+	})
+
+	metric.Inc()
+	err := b.Build()
+	metric.Dec()
+
+	return err
+}
+
+func (q *BuildQueue) runExclusiveBuild(b Builder) {
+	uuid := b.GetUUID()
+
+	for {
+		currentBuild := q.nominateBuild(uuid)
+		if currentBuild == nil {
+			return
+		}
+
+		q.exclusiveBuilds.waitForBuildCondition(q)
+		err := q.build(b)
+		q.exclusiveBuilds.finish(q)
+
+		currentBuild.reportStatus(q, err == nil)
+
+		if !q.needToBuildAgain(uuid) {
+			return
+		}
+
+		select {
+		case <-q.stopWorker:
+			return
+		default:
+		}
+	}
+}
+
+// PreemptExclusive enqueues a build at the front of the queue and provides
+// exclusive build access. All other builds enqueued via Enqueue()
+// PreemptExclusive() have to finish before the exclusive build is executed.
+// The exclusive build will then be the only build executed until it finishes.
+//
+// If an exclusive build for the same UUID is already enqueued but not yet
+// running, the build will be folded into the already scheduled build but both
+// notification channels will be notified.
+func (q *BuildQueue) PreemptExclusive(b Builder) BuildNotification {
```

Comment:
Same as for normal builds, I would return a read-only channel instead of a read-write `BuildNotification`.